frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/GitFetchSupportImpl.java

```diff
@@ -0,0 +1,463 @@
+          .getOrElse(Option.none());
+      if (option.isDefined() && option.get().equalsIgnoreCase("store")) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private List<SingleRemoteResult> waitForFetchTasks(List<FetchTask> tasks) {
+    var results = new ArrayList<SingleRemoteResult>();
+    for (var task : tasks) {
+      try {
+        results.add(task.future.get());
+      } catch (CancellationException | InterruptedException e) {
+        throw new ProcessCanceledException(e);
+      } catch (ExecutionException e) {
+        if (e.getCause() instanceof ProcessCanceledException) {
+          throw (ProcessCanceledException) e.getCause();
+        }
+        String msg = Option.of(e.getCause())
+            .flatMap(c -> Option.of(c.getMessage()))
+            .getOrElse("Error");
+        results.add(new SingleRemoteResult(task.repository, task.remote, msg, List.empty()));
+        LOG.error("Task execution error: ${msg}");
+      }
+    }
+    return List.ofAll(results);
+  }
+
+  private RepoResult mergeRepoResults(@Nullable RepoResult firstResult, SingleRemoteResult secondResult) {
+    if (firstResult == null) {
+      return new RepoResult(HashMap.of(secondResult.remote, secondResult));
+    } else {
+      var results = HashMap.of(secondResult.remote, secondResult).merge(firstResult.results);
+      return new RepoResult(results);
+    }
+  }
+
+  private <T> T withIndicator(Supplier<T> operation) {
+    var indicator = progressManager.getProgressIndicator();
+    var prevText = "";
+    if (indicator != null) {
+      prevText = indicator.getText();
+      indicator.setText("Fetching...");
+    }
+    try {
+      return operation.get();
+    } finally {
+      if (indicator != null) {
+        indicator.setText(prevText);
+      }
+    }
+  }
+
+  @Nullable
+  @Override
+  public GitRemote getDefaultRemoteToFetch(GitRepository repository) {
+    var remotes = repository.getRemotes();
+    if (remotes.isEmpty()) {
+      return null;
+    } else if (remotes.size() == 1) {
+      return remotes.iterator().next();
+    } else {
+      // this emulates behavior of the native `git fetch`:
+      // if current branch doesn't give a hint, then return "origin"; if there is no "origin", don't guess and fail
+      return Option.of(repository.getCurrentBranch())
+          .flatMap(b -> Option.of(b.findTrackedBranch(repository)))
+          .map(t -> t.getRemote())
+          .flatMap(r -> Option.of(findRemoteByName(repository, GitRemote.ORIGIN)))
+          .getOrNull();
+    }
+  }
+
+  private SingleRemoteResult doFetch(GitRepository repository, GitRemote remote, @Nullable String refspec,
+      GitAuthenticationGate authenticationGate) {
+    var recurseSubmodules = "--recurse-submodules=no";
+
+    // By default git fetch refuses to update the head which corresponds to the current branch.
+    // This flag disables the check.
+    var updateHeadOk = "--update-head-ok";
```

Comment:
As mentioned on Slack, document the minimum required version of git for our plugin in README.md

Reply:
_Slack discussion summary_

This option is available since 0.99. 
No documentation needed.